core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -97,6 +97,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         case requestId => throw new KafkaException("Unknown api code " + requestId)
       }
     } catch {
+      case fee: FatalExitError => throw fee
```

Comment:
Thanks for the explanation @maysamyabandeh and for your contribution. It's our bad for being slow on this one, no worries at all about not having time at the moment.

I'm thinking of making the minimal change required to fix the deadlock on shutdown and then leave additional changes for a subsequent PR. Also, I'm thinking we don't need an additional thread in the shutdown hook if we ensure that the latch in `ShutdownableThread` is open before we call `System.exit` (which seems safe). Let me know if you see problems with this approach (if you have a chance).